### ISSUE TITLE
Side-effect-free and/or immutable data on entities

### ISSUE THREAD

**Issue by joshua-berry-ntnx** [nutanix/papiea#432]
tl;dr: Not all properties associated with an entity need to have side-effects when they are changed, and not all properties on an entity should be mutable.

Today, Papiea has spec-only entities, where the action to resolve a diff between spec and status is always a no-op (i.e. status always becomes spec with no side-effects). However, I think there are situations where it would be useful to have "side-effect-free" properties on entities, as well as properties which are immutable. (I'm clubbing these two together because I'm not sure if they're the same or different feature requests. :) )

There are various axes (axises? axi? axia? axen?...English is weird) on which we can categorize properties:

* Present in spec, status or both? i.e.
** Is it observed (status), defined (spec), or both?
** Does it have side-effects when changed?
* Is it optional or required?
* Is it mutable or immutable?
** (Optional/immutable might mean it can only be sent once)

I'm not sure yet how these should be represented in Papiea, but opening an issue here as a placeholder for further discussion (and so I don't lose track of it).

**Comment by Emulebest** [comment]
Could you make a real-world example where this would be necessary. I didn't quite get the concept, even though I am familiar with the whole immutability stuff

**Comment by joshua-berry-ntnx** [comment]
Sure, for immutability, a simple example would be you have a machine entity, and that entity has a `type` field representing the type of machine this is. The machine type will never change (and indeed, the provider is not set up to deal with changes to a machine's type), so allowing it to be changed might result in undefined/incorrect behavior. So blocking changes to the machine-type field isn't just a security concern, it's also correctness—_nobody_ should change it, not even the provider itself.

For "side-effect-free" properties, an example might be a machine's loan information—say we loan a machine out to a customer, and keep some property about when the loan starts.  This information is not significant to the machine itself in any way, it's just a bit of administrative bookkeeping to satisfy business requirements. So the distinction between spec and status is meaningless. Obviously the loan start time will change as the machine is loaned to different customers, but there's not really an "intent" here, just a piece of data.

I think "optional vs required" is pretty straightforward, although it has interesting interactions with immutability—what is an "optional but immutable" property? Is it a property that can be set only once and never changed, or can it _only_ be set at entity creation time?

**Comment by Emulebest** [comment]
For side-effect-free isn't `status-only` portion already taking its place? There is not an intent there since it doesn't have a conforming spec field. The immutability part is nice but I think that there still should be a way to change the immutable field since even the internals of `machine type` could change and we could swap types on the fly. Immutable is immutable except when it is not :). Maybe having an error suppressed by `force` parameter on changing those fields would be sufficient.
As for the optionals... don't we have those already or they should function differently?

**Comment by joshua-berry-ntnx** [comment]
> For side-effect-free isn't status-only portion already taking its place? There is not an intent there since it doesn't have a conforming spec field.

`status-only` is different—only providers can change it (so is more akin to immutability); for "side-effect-free" properties, I'm asking for something clients can set. (And yes, I know this can be hacked by writing procedures.)

> The immutability part is nice but I think that there still should be a way to change the immutable field since even the internals of `machine type` could change and we could swap types on the fly. Immutable is immutable except when it is not :). Maybe having an error suppressed by `force` parameter on changing those fields would be sufficient.

No, this is exactly what I don't want. Any change whatsoever to a machine's type _is a bug_ and should be _completely_ disallowed for both providers and clients outside of a constructor or similar. (And there are other examples of this too which I've run across, for example the CIDR and owner of a subnet, the name of a machine type, etc.)

Changing a property from immutable to mutable should be a schema change (as it is in most programming languages—think about the invariants that change when you add `mut` to a reference in Rust, or `mutable` to a C++ member variable ;) ).

> As for the optionals... don't we have those already or they should function differently?

Yes we do, I was just calling them out because I think there might be some interaction to consider between optionality and immutability.

**Comment by Emulebest** [comment]
> > For side-effect-free isn't status-only portion already taking its place? There is not an intent there since it doesn't have a conforming spec field.
> 
> `status-only` is different—only providers can change it (so is more akin to immutability); for "side-effect-free" properties, I'm asking for something clients can set. (And yes, I know this can be hacked by writing procedures.)
> 
> > The immutability part is nice but I think that there still should be a way to change the immutable field since even the internals of `machine type` could change and we could swap types on the fly. Immutable is immutable except when it is not :). Maybe having an error suppressed by `force` parameter on changing those fields would be sufficient.
> 
> No, this is exactly what I don't want. Any change whatsoever to a machine's type _is a bug_ and should be _completely_ disallowed for both providers and clients outside of a constructor or similar. (And there are other examples of this too which I've run across, for example the CIDR and owner of a subnet, the name of a machine type, etc.)
> 
> Changing a property from immutable to mutable should be a schema change (as it is in most programming languages—think about the invariants that change when you add `mut` to a reference in Rust, or `mutable` to a C++ member variable ;) ).
> 
> > As for the optionals... don't we have those already or they should function differently?
> 
> Yes we do, I was just calling them out because I think there might be some interaction to consider between optionality and immutability.

I get your idea about immutable fields NEVER being able to change. Don't get me wrong, the idea is great but is it going to be practical? Even in pure functional languages there is also an escape hatch to opt-in mutability, that's why I proposed to have `force` as a flag to change this if really needed. @shlomi-nx what do you think?

**Comment by shlomi-nx** [comment]
Let's discuss this over zoom.

I am not sure I understand the merits of no-op fields. If their purpose is to allow a user to simply retain information on an intentful entity, isn't it just a simple "on" handler? the reason to have it as a handler is because it will be the only way to check any form of correctness on it. However, we could always add a "default" handler by specifying some `x-papiea-info-only` field in the schema and have that field behave like a "spec-only" entity. 

Regarding immutable spec fields, yes, I agree. In essence, these fields should be marked as `readonly` in their schema, and only the constructor handler (`on-create`) triggered by `POST` would be allowed to receive them. They should be disallowed in PUT.

**Comment by joshua-berry-ntnx** [comment]
Was going back over old notes and there is one more item I missed in my original description, which should maybe be a separate issue: user-defined key/value pairs. For example, AWS has the notion of "tags", which are arbitrary key/value pairs associated with a resource.  AWS itself ignores these; they are only present for user convenience. We might want something similar as well.
